src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -894,6 +894,38 @@ public T WithB2CAuthority(string authorityUri)
             return (T)this;
         }
 
+        /// <summary>
+        /// Adds a known authority corresponding to a generic OpenIdConnect Provider (OP).
+        /// See https://openid.net/specs/openid-connect-core-1_0.html#Terminology
+        /// </summary>
+        /// <param name="authorityUri">OpenIdConnect authority</param>
+        /// <param name="discoveryDocumentAddress">Address of the discovery endpoint. See https://openid.net/specs/openid-connect-discovery-1_0.html</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        public T WithGenericAuthority(string authorityUri, string discoveryDocumentAddress = null)
+        {
+            var authorityInfo = AuthorityInfo.FromGenericAuthority(authorityUri, discoveryDocumentAddress);
+            Config.Authority = GenericAuthority.CreateAuthority(authorityInfo);
+            return (T)this;
+        }
+        
+        /// <summary>
+        /// Configure a <see cref="IDocumentRetriever"/> to retrieve the discovery metadata from the OpenIdConnect Provider.
+        /// </summary>
+        /// <param name="documentRetriever"><see cref="IDocumentRetriever"/> instance responsible for retrieving the discovery metadata</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        /// <exception cref="NotSupportedException"> is thrown when not used with <see cref="WithGenericAuthority"/>.</exception>
+        public T WithDocumentRetriever(IDocumentRetriever documentRetriever)
```

Comment:
I don't think we want to expose this kind of extensibility at first, i.e. nobody asked for it. I think we should hide this for now?

Reply:
When working behind enterprise http proxy you need to be able to customize the underlying `HttpClient`. Exposing `IDocumentRetriever` allow one to configure this providing an `HttpDocumentRetriever` instance with its own HttpClient (please see https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httpdocumentretriever?view=msal-web-dotnet-latest#constructors). I could have exposed `WithHttpClient` to align with `OpenIdConnectConfigurationRetriever.Task<OpenIdConnectConfiguration> GetAsync(string address, HttpClient httpClient, CancellationToken cancel)` (please see https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.openidconnect.openidconnectconfigurationretriever.getasync?view=msal-web-dotnet-latest#overloads) but `IDocumentRetriever` is more generic and let one configure MSAL with offline metadata document by providing a `FileDocumentRetriever`.

Reply:
removed